File: backend/app/core/skills/registry.py
import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Dict, Iterable, List, Optional


@dataclass
class SkillSpec:
    id: str
    name: str
    description: str
    triggers: List[str]
    input_schema: Dict[str, Any]
    output_schema: Dict[str, Any]
    required_tools: List[str]
    default_model_role: str
    prompt: str = ""

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": self.id,
            "name": self.name,
            "description": self.description,
            "triggers": list(self.triggers),
            "input_schema": dict(self.input_schema),
            "output_schema": dict(self.output_schema),
            "required_tools": list(self.required_tools),
            "default_model_role": self.default_model_role,
            "prompt": self.prompt,
        }


class SkillRegistry:
    def __init__(self, skills: Optional[Iterable[SkillSpec]] = None) -> None:
        self._skills: Dict[str, SkillSpec] = {}
        for skill in skills or []:
            self.register(skill)

    def register(self, skill: SkillSpec) -> None:
        if not skill.id:
            raise ValueError("Skill id is required.")
        self._skills[skill.id] = skill

    def list(self) -> List[SkillSpec]:
        return sorted(self._skills.values(), key=lambda item: item.id)

    def get(self, skill_id: str) -> SkillSpec:
        try:
            return self._skills[skill_id]
        except KeyError as exc:
            raise KeyError(skill_id) from exc

    def select(self, text: str, methods: Optional[Iterable[str]] = None) -> List[SkillSpec]:
        q = (text or "").lower()
        selected: List[SkillSpec] = []
        method_text = " ".join(methods or []).lower()
        for skill in self.list():
            haystack = q + " " + method_text
            if any(trigger.lower() in haystack for trigger in skill.triggers):
                selected.append(skill)
        if not selected and self._skills:
            selected.append(self._skills["research_literature"])
        return selected


def create_default_skill_registry(extra_roots: Optional[Iterable[Path]] = None) -> SkillRegistry:
    registry = SkillRegistry(default_skill_specs())
    for root in default_skill_roots(extra_roots):
        load_skill_root(registry, root)
    return registry


def default_skill_roots(extra_roots: Optional[Iterable[Path]]) -> List[Path]:
    roots = []
    repo_root = Path(__file__).resolve().parents[4]
    roots.append(repo_root / "skills")
    roots.extend(Path(item) for item in (extra_roots or []))
    return roots


def load_skill_root(registry: SkillRegistry, root: Path) -> None:
    if not root.exists():
        return
    for skill_json in root.glob("*/skill.json"):
        skill = load_skill(skill_json)
        registry.register(skill)


def load_skill(skill_json: Path) -> SkillSpec:
    data = json.loads(skill_json.read_text(encoding="utf-8-sig"))
    prompt_path = skill_json.parent / "prompt.md"
    prompt = prompt_path.read_text(encoding="utf-8") if prompt_path.exists() else ""
    return SkillSpec(
        id=str(data["id"]),
        name=str(data.get("name") or data["id"]),
        description=str(data.get("description") or ""),
        triggers=[str(item) for item in data.get("triggers", [])],
        input_schema=dict(data.get("input_schema") or {"type": "object"}),
        output_schema=dict(data.get("output_schema") or {"type": "object"}),
        required_tools=[str(item) for item in data.get("required_tools", [])],
        default_model_role=str(data.get("default_model_role") or "executor"),
        prompt=prompt,
    )


def default_skill_specs() -> List[SkillSpec]:
    common_input = {"type": "object", "properties": {"question": {"type": "string"}}}
    common_output = {"type": "object", "properties": {"evidenceIds": {"type": "array"}}}
    return [
        SkillSpec(
            id="research_literature",
            name="Research Literature",
            description="Search and summarize literature-grade evidence for a research question.",
            triggers=["literature", "paper", "research", "review", "文献", "论文", "检索", "研究"],
            input_schema=common_input,
            output_schema=common_output,
            required_tools=["literature.local_search", "mcp.call"],
            default_model_role="executor",
            prompt="Collect source-grounded evidence and keep every claim traceable.",
        ),
        SkillSpec(
            id="paper_analysis",
            name="Paper Analysis",
            description="Analyze papers, claims, methods, limitations, and citation support.",
            triggers=["paper", "citation", "doi", "pdf", "论文", "引用", "方法"],
            input_schema=common_input,
            output_schema=common_output,
            required_tools=["literature.local_search"],
            default_model_role="executor",
            prompt="Extract claims, methods, assumptions, limitations, and evidence links.",
        ),
        SkillSpec(
            id="experiment_design",
            name="Experiment Design",
            description="Design reproducible experiments, checks, and analysis artifacts.",
            triggers=["experiment", "analysis", "simulate", "python", "实验", "分析", "仿真"],
            input_schema=common_input,
            output_schema={"type": "object", "properties": {"artifacts": {"type": "array"}}},
            required_tools=["python.sandbox"],
            default_model_role="executor",
            prompt="Design a reproducible experiment or analysis path with explicit assumptions.",
        ),
    ]

File: genesis/skills/registry.ts
import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import os from "node:os";
import path from "node:path";
import type { MCPTool, ModelRole, Skill } from "../types/research.ts";

export type SkillAuditIssue = {
  skill_id: string;
  severity: "warning" | "error";
  message: string;
};

const LEGACY_TOOL_ALIASES: Record<string, string[]> = {
  "python.sandbox": ["runtime.python"],
  "literature.local_search": ["literature.search"],
  "mcp.call": ["browser.validate"]
};

export class SkillRegistry {
  private readonly skills = new Map<string, Skill>();
  private readonly aliases = new Map<string, string>();

  constructor(initialSkills: Skill[] = defaultSkills(), options: { skillRoot?: string; loadDirectory?: boolean } = {}) {
    for (const skill of initialSkills) {
      this.register(skill);
    }
    if (options.loadDirectory ?? true) {
      for (const skill of loadSkillsFromDirectory(options.skillRoot ?? defaultSkillRoot())) {
        this.register(skill);
      }
    }
  }

  register(skill: Skill): void {
    this.skills.set(skill.id, skill);
    this.aliases.set(skill.name, skill.id);
  }

  get(name: string): Skill {
    const id = this.skills.has(name) ? name : this.aliases.get(name);
    const skill = id ? this.skills.get(id) : undefined;
    if (!skill) {
      throw new Error(`Unknown skill: ${name}`);
    }
    return skill;
  }

  list(): Skill[] {
    return Array.from(this.skills.values());
  }

  audit(availableTools: Iterable<string | MCPTool> = []): SkillAuditIssue[] {
    const toolNames = new Set(Array.from(availableTools, (tool) => typeof tool === "string" ? tool : tool.name));
    const issues: SkillAuditIssue[] = [];
    for (const skill of this.list()) {
      if (!skill.system_prompt.trim()) {
        issues.push({ skill_id: skill.id, severity: "error", message: "system_prompt is empty." });
      }
      if (!skill.workflow.length) {
        issues.push({ skill_id: skill.id, severity: "warning", message: "workflow is empty." });
      }
      for (const tool of [...skill.required_tools, ...skill.tool_policy.allowed_tools]) {
        if (LEGACY_TOOL_ALIASES[tool]) {
          issues.push({ skill_id: skill.id, severity: "error", message: `legacy tool alias remains: ${tool}` });
        }
        if (toolNames.size > 0 && !toolNames.has(tool) && !tool.endsWith(".*")) {
          issues.push({ skill_id: skill.id, severity: "warning", message: `tool is not registered by default MCP registry: ${tool}` });
        }
      }
    }
    return issues;
  }

  selectForIdea(idea: string): Skill[] {
    const q = idea.toLowerCase();
    const tokens = q.split(/[^a-z0-9_]+/).filter((token) => token.length > 3);
    const selected = this.list().filter((skill) => {
      const haystack = [skill.id, skill.name, skill.description, skill.system_prompt, ...skill.triggers, ...skill.workflow].join(" ").toLowerCase();
      return skill.triggers.some((trigger) => q.includes(trigger.toLowerCase())) || tokens.some((token) => haystack.includes(token));
    });
    return selected.length ? selected : [this.get("research_skill")];
  }
}

export function loadSkillsFromDirectory(root = defaultSkillRoot()): Skill[] {
  if (!existsSync(root)) {
    return [];
  }
  const skills: Skill[] = [];
  for (const entry of readdirSync(root)) {
    const dir = path.join(root, entry);
    if (!statSync(dir).isDirectory()) {
      continue;
    }
    const specPath = path.join(dir, "skill.json");
    if (!existsSync(specPath)) {
      continue;
    }
    try {
      const raw = JSON.parse(readFileSync(specPath, "utf8")) as Record<string, unknown>;
      const promptPath = path.join(dir, "prompt.md");
      const prompt = existsSync(promptPath) ? readFileSync(promptPath, "utf8").trim() : String(raw.system_prompt ?? raw.description ?? "");
      skills.push(normalizeSkillSpec(raw, prompt));
    } catch {
      continue;
    }
  }
  return skills;
}

function normalizeSkillSpec(raw: Record<string, unknown>, prompt: string): Skill {
  const id = stringValue(raw.id, stringValue(raw.name, "unknown_skill")).replace(/\s+/g, "_");
  const requiredTools = canonicalToolList(stringArray(raw.required_tools));
  const workflow = stringArray(raw.workflow, stringArray(raw.triggers));
  return {
    id,
    name: stringValue(raw.name, id).replace(/\s+/g, "_"),
    description: stringValue(raw.description, prompt || id),
    triggers: stringArray(raw.triggers),
    input_schema: raw.input_schema ?? { type: "object" },
    output_schema: raw.output_schema ?? { type: "object" },
    required_tools: requiredTools,
    default_model_role: normalizeModelRole(stringValue(raw.default_model_role, "execution")),
    system_prompt: prompt || stringValue(raw.system_prompt, stringValue(raw.description, id)),
    tool_policy: normalizeToolPolicy(raw.tool_policy, requiredTools),
    workflow
  };
}

function normalizeToolPolicy(value: unknown, requiredTools: string[]): Skill["tool_policy"] {
  if (value && typeof value === "object" && !Array.isArray(value)) {
    const record = value as Record<string, unknown>;
    return {
      allowed_tools: canonicalToolList(stringArray(record.allowed_tools, requiredTools)),
      disallowed_tools: canonicalToolList(stringArray(record.disallowed_tools))
    };
  }
  return {
    allowed_tools: requiredTools.length ? requiredTools : ["literature.search", "browser.validate"],
    disallowed_tools: []
  };
}

function canonicalToolList(tools: string[]): string[] {
  return Array.from(new Set(tools.flatMap((tool) => LEGACY_TOOL_ALIASES[tool] ?? [tool])));
}
function normalizeModelRole(value: string): ModelRole {
  if (value === "planner") {
    return "planning";
  }
  if (value === "executor") {
    return "execution";
  }
  if (value === "planning" || value === "execution" || value === "critic" || value === "synthesizer") {
    return value;
  }
  return "execution";
}

function stringValue(value: unknown, fallback: string): string {
  return typeof value === "string" && value.trim() ? value.trim() : fallback;
}

function stringArray(value: unknown, fallback: string[] = []): string[] {
  return Array.isArray(value) ? value.filter((item): item is string => typeof item === "string") : fallback;
}

export function defaultSkillRoot(): string {
  return process.env.GENESIS_SKILLS_DIR || path.join(process.env.GENESIS_HOME || path.join(os.homedir(), ".genesis"), "skills");
}

export function defaultSkills(): Skill[] {
  return [
    {
      id: "research_skill",
      name: "research_skill",
      description: "Plan and execute source-grounded scientific research with explicit evidence.",
      triggers: ["research", "evidence", "hypothesis", "study", "literature"],
      input_schema: { type: "object" },
      output_schema: { type: "object" },
      required_tools: ["literature.search", "browser.validate", "dataset.lookup"],
      default_model_role: "planning",
      system_prompt: "Plan and execute source-grounded scientific research with explicit evidence.",
      tool_policy: {
        allowed_tools: ["literature.search", "browser.validate", "dataset.lookup", "github.repo_exploration"],
        disallowed_tools: ["github.change_execution"]
      },
      workflow: ["decompose idea", "collect evidence", "map evidence to claims", "synthesize report"]
    },
    {
      id: "paper_analysis_skill",
      name: "paper_analysis_skill",
      description: "Analyze papers, methods, claims, limitations, and citation support.",
      triggers: ["paper", "citation", "doi", "pdf", "literature"],
      input_schema: { type: "object" },
      output_schema: { type: "object" },
      required_tools: ["literature.search", "browser.validate", "dataset.lookup"],
      default_model_role: "execution",
      system_prompt: "Analyze papers, methods, claims, limitations, and citation support.",
      tool_policy: {
        allowed_tools: ["literature.search", "browser.validate", "dataset.lookup"],
        disallowed_tools: ["github.change_execution"]
      },
      workflow: ["identify claims", "inspect methods", "extract evidence", "flag limitations"]
    },
    {
      id: "coding_skill",
      name: "coding_skill",
      description: "Use code execution as a research instrument, not as automation workflow.",
      triggers: ["code", "repo", "github", "experiment", "simulate", "python"],
      input_schema: { type: "object" },
      output_schema: { type: "object" },
      required_tools: ["runtime.python", "github.code_understanding", "github.repo_exploration"],
      default_model_role: "execution",
      system_prompt: "Use code execution as a research instrument, not as automation workflow.",
      tool_policy: {
        allowed_tools: ["runtime.python", "github.code_understanding", "github.repo_exploration", "dataset.lookup"],
        disallowed_tools: []
      },
      workflow: ["inspect code", "design experiment", "run minimal reproducible analysis", "record artifacts"]
    },
    {
      id: "debugging_skill",
      name: "debugging_skill",
      description: "Diagnose failures as evidence for research hypotheses.",
      triggers: ["debug", "failure", "ci", "bug", "trace"],
      input_schema: { type: "object" },
      output_schema: { type: "object" },
      required_tools: ["github.ci_diagnosis", "github.code_understanding", "runtime.python"],
      default_model_role: "critic",
      system_prompt: "Diagnose failures as evidence for research hypotheses.",
      tool_policy: {
        allowed_tools: ["github.ci_diagnosis", "github.code_understanding", "runtime.python", "browser.validate"],
        disallowed_tools: ["github.change_execution"]
      },
      workflow: ["debug github ci", "collect logs", "classify failure", "test hypothesis", "summarize contradiction"]
    }
  ];
}
